feat: anonymous access and API features

davidjoy opened the pull request:
Features:

**requireAuthenticatedUser** - Turns on automatic login redirection for unauthenticated users.  When turned off, allows for consumers of frontend-base to make public MFEs.
**hydrateAuthenticatedUser** - A replacement for fetchUserAccount in frontend-auth, populates App.authenticatedUser with detailed user account information for the current user from the same API as fetchUserAccount.
**AuthenticatedRoute** - A react-router-dom compatible Component that performs a login redirect if a user visits the route while unauthenticated.
**Api** - A collection of API helper functions to ease API interactions.
**useAppEvent** - A React hook allowing functional components to manage subscriptions to App events.
**AppContext updates** - AppContext will now receive updates to authenticatedUser and config, allowing React applications to respond to those changes.
**LoginRedirect** - A React Component that, when rendered, will redirect to the login page.
**getAuthenticatedUserAccount** - An API method which loads account data for the authenticated user.  This is called internally by the hydrateAuthenticatedUser option.

Fixes:

Fixed a typo in the ErrorPage component.

BREAKING CHANGES:

**fetchUserAccount** - no longer supported.  Use hydrateAuthenticatedUser and AppContext instead.

Commit:
feat: anonymous access and API features

Features:

allowAnonymous - Turns off automatic login redirection for unauthenticated users, allowing for consumers of frontend-base to make public MFEs.
hydrateAuthenticatedUser - A replacement for fetchUserAccount in frontend-auth, populates App.authenticatedUser with detailed user account information for the current user from the same API as fetchUserAccount.
AuthenticatedRoute - A react-router-dom compatible Component that performs a login redirect if a user visits the route while unauthenticated.
Api - A collection of API helper functions to ease API interactions.
useAppEvent - A React hook allowing functional components to manage subscriptions to App events.
AppContext updates - AppContext will now receive updates to authenticatedUser and config, allowing React applications to respond to those changes.
LoginRedirect - A React Component that, when rendered, will redirect to the login page.
getAuthenticatedUserAccount - An API method which loads account data for the authenticated user.  This is called internally by the hydrateAuthenticatedUser option.

Fixes:

Fixed a typo in the ErrorPage component.

Commit:
docs: updating API doc with the majority of new features

Will document Api.js functions in a subsequent PR.

Also fixing the dev server port.

abutterworth commented on docs/API.rst:
```diff
@@ -64,11 +64,20 @@ error phase.
-keys:
+The ``App.initialize`` method takes an options object with the following possible keys:
+
+allowAnonymous
```
Is this something that could be supplied as a prop to <AppProvider>? Does it make more sense on init or there I wonder. Will continue to think about as I read. Maybe:
```
<AppProvider allowAnonymous>

</AppProvider>
```

abutterworth commented on src/Api.js:
```diff
@@ -0,0 +1,82 @@
+  return modifyObjectKeys(object, transformer);
+};
+
+export function formatKnownError(error) {
```
Maybe getting the word `user` facing errors in the name would help make this clearer

davidjoy replied:
I wasn't 100% on the name of this - they're things from the server we know how to handle and display to the user.  I think your comment about changing it to something like `formatUserFacingError` or something like that would make this clearer.

davidjoy committed:
Update API.rst

abutterworth commented on src/Api.js:
```diff
@@ -0,0 +1,82 @@
+  );
+}
+
+export const request = async (type, ...args) => {
```
Could we leverage axios interceptors instead?

davidjoy replied:
This code is being removed - we decided not to include it since we're not convinced it's the right shape.

davidjoy commented on docs/API.rst:
```diff
@@ -277,17 +317,35 @@ the following data structure:
+
+In the above example, an anonmyous/unauthenticated user navigating to /authenticated will be redirected to the login page.
+
+``LoginRedirect``
```
Because this uses AppContext, I don't think I can.  Think I'll leave it for now...

davidjoy commented on src/index.js:
```diff
@@ -10,15 +10,35 @@ export {
+  isKnownError,
+  request,
+  get,
+  del,
```
🌞

Commit:
fix: review feedback

BREAKING CHANGE: Renaming allowAnonymous to requireAuthenticatedUser and turning it off by default.  This breaks current MFEs, which will now be required to turn on requireAuthenticatedUser.

Commit:
fix: removing support for fetchUserAccount

BREAKING CHANGE: fetchUserAccount is no longer supported.  Use hydrateAuthenticatedUser instead, along with AppContext.

davidjoy commented on src/index.js:
```diff
@@ -30,3 +50,11 @@ export { default as error } from './handlers/error';
+export { default as validateConfig } from './data/validateConfig';
+export { default as env } from './data/env';
+export { useAppEvent } from './data/hooks';
+export { fetchUserAccount } from './frontendAuthWrapper';
```
Rip't
